### PR TITLE
feat: migrates Container to TypeScript; Container without max-width on docs site [FC-0062]

### PR DESCRIPTION
**Pull request by rpenido**
## Description

This PR allows the `size` property of the `Container` to be `undefined`, removing the maximum width in this case

### Deploy Preview

https://deploy-preview-3216--paragon-openedx.netlify.app/components/container/

## More information
Part of:
- https://github.com/openedx/frontend-app-course-authoring/issues/1221

## Merge Checklist

* [x] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [x] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [x] Were your changes tested in the `example` app?
* [x] Is there adequate test coverage for your changes?
* [x] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please request an a11y review for the PR in the [#wg-paragon](https://openedx.slack.com/archives/C02NR285KV4) Open edX Slack channel.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

___
Private ref: [FAL-3820](https://tasks.opencraft.com/browse/FAL-3820)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @rpenido!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/paragon-working-group`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: src/Container/index.tsx
Comment (pomegranited) [inline src/Container/index.tsx:32]: nit: Do we need these lines when these properties are overridden by the prop defaults below?

I'm more concerned with the `fluid` one, since I'm unsure how overriding these boolean values works in Typescript, and don't want to break anyone else's UI.

```suggestion
```

File: src/Container/Container.test.tsx
Comment (pomegranited) [inline src/Container/Container.test.tsx:35]: I'd rather use an explicit `size="full"` than assume that omitting the size means "full width".. Omitting a size should make the container fit its contents, just like a `<div>` without a `max-width`.
I think the bootstrap class `mw-100` does this?

**Review by pomegranited — CHANGES_REQUESTED** [review]
@rpenido Some comments for you -- I still need to test this too.

And in future, could you separate out converting from JSX -> TSX into a separate commit? That would help me see what's changed :)

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx:32]: You're right. We don't need this.
Removed here: e96c4608b57a880ae31f4427c9c4e4791cc18ef0

As information only, because the spread `{...props}` was after these props, the spread would override it. The boolean `fluid` is a syntax sugar for `fluid={true}`. This is the transpiled JS code:

image

File: src/Container/README.md
Comment (rpenido) [reply src/Container/README.md:23]: Fixed: e96c4608b57a880ae31f4427c9c4e4791cc18ef0

File: src/Container/Container.test.tsx
Comment (rpenido) [reply src/Container/Container.test.tsx:35]: I agree with you and tried this approach first.

The problem is that the current component allows the `size` props to be omitted, and it behaves like "full width". You can check it here, removing the `size` from one example: https://paragon-openedx.netlify.app/components/container/

We have many containers in our code base without `size` that probably rely on this, so I don't think we can change that.
So I pushed back the `size={full}` option.

Comment (rpenido) [reply src/Container/Container.test.tsx:35]: I tried to update the API docs here: bedcb9b914d5eb47518f15d9e9d34841bfb37e45

**Comment by adamstankiewicz** [comment]
Thanks for the contribution! Just as a heads up, I believe the primary change here is the TypeScript migration and adding the docs site example for `Container` without a `max-width`. The current implementation of `Container` supports the intended behavior today when omitting the `size` prop: [Playground Example](https://paragon-openedx.netlify.app/playground/?code=N4Igxg9gJgpiBcIA8UCWA3ABGANgQwGcCA5PAWxgF4AdEAIwHMBaHVBgCwBcmAmABj6YADgE8mAVloA%2BagDtMmJAGVOeMAGtMDPEMrAALADpxAXxnyFigMIRZq1LJgAnbPiKkKNes1YdO0gAkYHBwITAB3CCccKCQAehs7PAdnKUxgOIAqTGIAeQAVAFF4TFkwsjwADyZw1ChOdgj2GHkAAwJUAC8YVsxUAlKITmFCAhgoTEy4kzlLa1t7RxcO7q9KglpXUY8qWkYWNi5A4NCIqJj4xMXU2cskK%2BSlzBXdkAIyTdxt8lf93yOQFIgiEwpForEEgtHjcLAp7lCUssuq8yFBPm4SD8vH9Dv5AcDTmCLpCkojzHN4aSni8vDgGOjvp49j5cccQWdwZcEUtyYo4io1OpzPE0OgpCAADQgBowCgEBAAbRAECELXGlRAAF0pbV6ux5fAFQBGfQCTUmIA).

The first instance of `Container` is the above example renders with `max-width: none;`, when `size` is omitted or an invalid `size` is passed:

image

File: src/Container/index.tsx
Comment (adamstankiewicz) [inline src/Container/index.tsx:1]: [optional] With a recent update to `@edx/eslint-config`, we might be able to upgrade to v4.2.0 in this repo and remove the need for default params in this file without disabling the ESLint rule 👀 https://github.com/openedx/eslint-config/pull/156

While this repo is currently using v3 of `@edx/eslint-config`, the v4 breaking changes don't seem to apply to this repo at the moment, per the release notes, so the upgrade should be fairly trivial: 

> Drops support for `eslint` ^6.8.0; `@edx/eslint-config` must now be used with at least `eslint` versions ^7.32.0 or ^8.2.0.

File: src/Container/index.tsx
Comment (adamstankiewicz) [inline src/Container/index.tsx:23]: Is the `Deprecated` property necessary here? `Container` doesn't have a deprecated sub-component (e.g., `Container.Deprecated`) as `Button.Deprecated` does.

File: src/Container/index.tsx
Comment (adamstankiewicz) [inline src/Container/index.tsx:25]: Could `Container` be used with a non-`div` element, e.g. with `as="span"`? I wonder if this should be more generic, e.g. `HTMLElement` or even `Element`.

File: src/Container/index.tsx
Comment (adamstankiewicz) [inline src/Container/index.tsx:51]: I believe the `undefined` is implied by the lack of `.isRequired` on the prop type definition?

File: www/src/context/SettingsContext.tsx
Comment (adamstankiewicz) [inline www/src/context/SettingsContext.tsx:15]: Thanks for adding the `ContainerSize` type here, too! 🙌

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx]: No, it's not needed. Thank you for catching this.
Fixed here: 641dd15975555d4035920afe71ba2bb8a18c3245

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx]: Yes! I changed it to `Element` as our property of interest (`className`) exists.
Here: 7e0fb5c8fffafed9b8fefa434a0306263a9312a4

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx]: Agreed. It was there to show in the docs, but it isn't necessary with the docstring change.
Fixed: 7a03d831af58a60e33492af763b3651aeb1fb2b

**Comment by rpenido** [comment]
> Just as a heads up, I believe the primary change here is the TypeScript migration and adding the docs site example for `Container` without a `max-width`.

Absolutely. Do you think I should change anything in the PR to make it clear?

**Comment by adamstankiewicz** [comment]
> Absolutely. Do you think I should change anything in the PR to make it clear?

I might just recommend amending your PR title / commit message to more closely reflect the TypeScript migration and the additional example on the docs site, e.g.

```
feat: migrates Container to TypeScript; Container without max-width on docs site
```

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx:1]: @adamstankiewicz  I tried to update but got the following error on linting (which prevents the commit).
Did you already experienced this error before?

```
> @openedx/paragon@20.18.1 lint
> npm run stylelint && eslint --ext .js --ext .jsx --ext .ts --ext .tsx . && npm run lint --workspaces --if-present


> @openedx/paragon@20.18.1 stylelint
> stylelint "src/**/*.scss" "scss/**/*.scss" "www/src/**/*.scss" --config .stylelintrc.json


Oops! Something went wrong! :(

ESLint: 8.18.0

Error: Error while loading rule '@typescript-eslint/naming-convention': You have used a rule which requires parserServices to be generated. You must therefore provide a value for the "parserOptions.project" property for @typescript-eslint/parser.
Occurred while linting /home/rpenido/Projects/opencraft/paragon/.eslintrc.js
...
```

Comment (adamstankiewicz) [reply src/Container/index.tsx:1]: Hmm, I have not. Thanks for flagging. I am able to reproduce the issue on my end as well. Unless you want to investigate the error further, feel free to defer on the upgrade and continue with your existing solution using default props and the disabled ESLint rule. If you do defer on the investigation/resolution, do you mind filing a new Paragon issue to document the error?

**Comment by adamstankiewicz** [comment]
@rpenido When this merges, to ensure it gets published to NPM via `semantic-release`, the commit type prefix will be `feat`. `semantic-release`, by default (as configured), only triggers NPM releases for `fix`, `feat`, and `perf`. Given your contribution is exposing new TypeScript types to consumers, I want to be sure merging triggers a new release `feat` :)

**Comment by rpenido** [comment]
Thank you, @adamstankiewicz! I updated the PR title again.
When you and @pomegranited think this is good for merge, just ping me, and I can squash it (if needed) to ensure we have the correct commit message.

**Review by adamstankiewicz — APPROVED** [review]

**Comment by adamstankiewicz** [comment]
LGTM! If either you or @pomegranited don't have the ability to merge, let me know and I'll happily merge it, too :)

File: src/Container/Container.test.tsx
Comment (pomegranited) [reply src/Container/Container.test.tsx:35]: > The problem is that the current component allows the size props to be omitted, and it behaves like "full width". You can check it here, removing the size from one example: https://paragon-openedx.netlify.app/components/container/

Passing in `size=""` gives us the fullscreen behaviour you're adding here.. so do we need this change?

**Comment by pomegranited** [comment]
Thanks @adamstankiewicz could you merge? We don't have access.

**Review by pomegranited — APPROVED** [review]
:+1:

- [x] I tested this while testing https://github.com/openedx/frontend-app-course-authoring/pull/1258#pullrequestreview-2291515898
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

File: src/Container/index.tsx
Comment (rpenido) [reply src/Container/index.tsx:1]: Issue created: https://github.com/openedx/paragon/issues/3230

**Pull request merged**
